datasource/src/main/scala/quasar/physical/s3/RequestSigning.scala

```diff
@@ -0,0 +1,201 @@
+/*
+ * Copyright 2014–2018 SlamData Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package quasar.physical.s3
+
+import slamdata.Predef._
+import java.net.URLEncoder
+import java.nio.charset.StandardCharsets
+import java.security.MessageDigest
+import java.time.LocalDateTime
+import java.time.format.DateTimeFormatter
+import javax.crypto.Mac
+import javax.crypto.spec.SecretKeySpec
+import cats.effect.Sync
+import cats.implicits._
+import fs2.Stream
+import org.http4s.headers.{Authorization, Date}
+import org.http4s.{Header, Headers, Method, Request, Uri}
+
+/**
+  * Extracted from aws4s: https://github.com/aws4s/aws4s
+  */
+
+/**
+  * Based on https://github.com/ticofab/aws-request-signer,
+  * inspired by: https://github.com/inreachventures/aws-signing-request-interceptor
+  */
+object RequestSigning {
+
+  private def sha256[F[_]: Sync](payload: Stream[F, Byte]): F[Array[Byte]] =
+    payload.chunks.compile.fold(MessageDigest.getInstance("SHA-256"))((md, chunk) => { md.update(chunk.toArray); md }).map(_.digest)
+
+  private def sha256(payload: Array[Byte]): Array[Byte] = {
+    val md: MessageDigest = MessageDigest.getInstance("SHA-256")
+    md.update(payload)
+    md.digest
+  }
+
+  private def base16(data: Array[Byte]): String = {
+    val BASE16MAP = List('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', 'a', 'b', 'c', 'd', 'e', 'f')
+    data.flatMap(byte => Array(BASE16MAP(byte >> 4 & 0xF), BASE16MAP(byte & 0xF))).mkString
+  }
+
+  private def renderCanonicalQueryString(queryParams: Map[String, String]): String =
+    queryParams.toSeq
+      .sortBy(_._1)
+      .map({ case (k, v) => k + "=" + URLEncoder.encode(v, StandardCharsets.UTF_8.toString) })
+      .mkString("&")
+
+  private def hmacSha256(data: String, key: Array[Byte]): Array[Byte] = {
+    val mac: Mac = Mac.getInstance("HmacSHA256")
+    mac.init(new SecretKeySpec(key, "HmacSHA256"))
```

Comment:
Can you make this string and `"SHA-256"` above constant `val`s in scope.

Reply:
sure!